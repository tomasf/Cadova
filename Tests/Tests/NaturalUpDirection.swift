import Testing
@testable import Cadova

struct NaturalUpDirectionTests {
    @Test func basics() async throws {
        try await Stack(.z, alignment: .center) {
            Cylinder(diameter: 1, height: 5)
            Cylinder(bottomDiameter: 2, topDiameter: 0, height: 2)
        }
        .readingEnvironment(\.naturalUpDirection) { arrow, globalUp in
            arrow.rotated(from: .up, to: globalUp ?? .up)
        }
        .translated(z: 7)
        .repeated(around: .x, count: 8)
        .definingNaturalUpDirection()
        .expectEquals(goldenFile: "naturalUpDirection")
    }
}
