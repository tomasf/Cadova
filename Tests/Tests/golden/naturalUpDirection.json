{
  "parts" : [
    {
      "defaultMaterial" : {
        "baseColor" : {
          "alpha" : 1,
          "blue" : 1,
          "green" : 1,
          "red" : 1
        }
      },
      "name" : "Model",
      "type" : "solid"
    },
    {
      "children" : [
        {
          "body" : {
            "children" : [
              {
                "body" : {
                  "kind" : "shape3D",
                  "primitive" : {
                    "cylinder" : {
                      "bottomRadius" : 0.5,
                      "height" : 5,
                      "segmentCount" : 20,
                      "topRadius" : 0.5
                    }
                  }
                },
                "kind" : "transform",
                "transform" : [
                  [
                    1,
                    0,
                    0,
                    0
                  ],
                  [
                    0,
                    1,
                    0,
                    0
                  ],
                  [
                    0,
                    0,
                    1,
                    0
                  ],
                  [
                    0,
                    0,
                    0,
                    1
                  ]
                ]
              },
              {
                "body" : {
                  "kind" : "shape3D",
                  "primitive" : {
                    "cylinder" : {
                      "bottomRadius" : 1,
                      "height" : 2,
                      "segmentCount" : 41,
                      "topRadius" : 0
                    }
                  }
                },
                "kind" : "transform",
                "transform" : [
                  [
                    1,
                    0,
                    0,
                    -0.001467099408129724
                  ],
                  [
                    0,
                    1,
                    0,
                    0
                  ],
                  [
                    0,
                    0,
                    1,
                    5
                  ],
                  [
                    0,
                    0,
                    0,
                    1
                  ]
                ]
              }
            ],
            "kind" : "lazyUnion"
          },
          "kind" : "transform",
          "transform" : [
            [
              1,
              0,
              0,
              0
            ],
            [
              0,
              1,
              0,
              0
            ],
            [
              0,
              0,
              1,
              7
            ],
            [
              0,
              0,
              0,
              1
            ]
          ]
        },
        {
          "body" : {
            "children" : [
              {
                "body" : {
                  "kind" : "shape3D",
                  "primitive" : {
                    "cylinder" : {
                      "bottomRadius" : 0.5,
                      "height" : 5,
                      "segmentCount" : 20,
                      "topRadius" : 0.5
                    }
                  }
                },
                "kind" : "transform",
                "transform" : [
                  [
                    1,
                    0,
                    0,
                    0
                  ],
                  [
                    0,
                    1,
                    0,
                    0
                  ],
                  [
                    0,
                    0,
                    1,
                    0
                  ],
                  [
                    0,
                    0,
                    0,
                    1
                  ]
                ]
              },
              {
                "body" : {
                  "kind" : "shape3D",
                  "primitive" : {
                    "cylinder" : {
                      "bottomRadius" : 1,
                      "height" : 2,
                      "segmentCount" : 41,
                      "topRadius" : 0
                    }
                  }
                },
                "kind" : "transform",
                "transform" : [
                  [
                    1,
                    0,
                    0,
                    -0.001467099408129724
                  ],
                  [
                    0,
                    1,
                    0,
                    0
                  ],
                  [
                    0,
                    0,
                    1,
                    5
                  ],
                  [
                    0,
                    0,
                    0,
                    1
                  ]
                ]
              }
            ],
            "kind" : "lazyUnion"
          },
          "kind" : "transform",
          "transform" : [
            [
              1,
              0,
              0,
              0
            ],
            [
              0,
              1,
              -3.1339616320458145e-16,
              -4.949747468305832
            ],
            [
              0,
              2.781780846884627e-16,
              1,
              4.949747468305833
            ],
            [
              0,
              0,
              0,
              1
            ]
          ]
        },
        {
          "body" : {
            "children" : [
              {
                "body" : {
                  "kind" : "shape3D",
                  "primitive" : {
                    "cylinder" : {
                      "bottomRadius" : 0.5,
                      "height" : 5,
                      "segmentCount" : 20,
                      "topRadius" : 0.5
                    }
                  }
                },
                "kind" : "transform",
                "transform" : [
                  [
                    1,
                    0,
                    0,
                    0
                  ],
                  [
                    0,
                    1,
                    0,
                    0
                  ],
                  [
                    0,
                    0,
                    1,
                    0
                  ],
                  [
                    0,
                    0,
                    0,
                    1
                  ]
                ]
              },
              {
                "body" : {
                  "kind" : "shape3D",
                  "primitive" : {
                    "cylinder" : {
                      "bottomRadius" : 1,
                      "height" : 2,
                      "segmentCount" : 41,
                      "topRadius" : 0
                    }
                  }
                },
                "kind" : "transform",
                "transform" : [
                  [
                    1,
                    0,
                    0,
                    -0.001467099408129724
                  ],
                  [
                    0,
                    1,
                    0,
                    0
                  ],
                  [
                    0,
                    0,
                    1,
                    5
                  ],
                  [
                    0,
                    0,
                    0,
                    1
                  ]
                ]
              }
            ],
            "kind" : "lazyUnion"
          },
          "kind" : "transform",
          "transform" : [
            [
              1,
              0,
              0,
              0
            ],
            [
              0,
              1,
              0,
              -7
            ],
            [
              0,
              0,
              1,
              4.286263797015736e-16
            ],
            [
              0,
              0,
              0,
              1
            ]
          ]
        },
        {
          "body" : {
            "children" : [
              {
                "body" : {
                  "kind" : "shape3D",
                  "primitive" : {
                    "cylinder" : {
                      "bottomRadius" : 0.5,
                      "height" : 5,
                      "segmentCount" : 20,
                      "topRadius" : 0.5
                    }
                  }
                },
                "kind" : "transform",
                "transform" : [
                  [
                    1,
                    0,
                    0,
                    0
                  ],
                  [
                    0,
                    1,
                    0,
                    0
                  ],
                  [
                    0,
                    0,
                    1,
                    0
                  ],
                  [
                    0,
                    0,
                    0,
                    1
                  ]
                ]
              },
              {
                "body" : {
                  "kind" : "shape3D",
                  "primitive" : {
                    "cylinder" : {
                      "bottomRadius" : 1,
                      "height" : 2,
                      "segmentCount" : 41,
                      "topRadius" : 0
                    }
                  }
                },
                "kind" : "transform",
                "transform" : [
                  [
                    1,
                    0,
                    0,
                    -0.001467099408129724
                  ],
                  [
                    0,
                    1,
                    0,
                    0
                  ],
                  [
                    0,
                    0,
                    1,
                    5
                  ],
                  [
                    0,
                    0,
                    0,
                    1
                  ]
                ]
              }
            ],
            "kind" : "lazyUnion"
          },
          "kind" : "transform",
          "transform" : [
            [
              1,
              0,
              0,
              0
            ],
            [
              0,
              0.9999999999999998,
              4.799296168983549e-16,
              -4.949747468305832
            ],
            [
              0,
              -4.3518733055684016e-16,
              0.9999999999999998,
              -4.949747468305832
            ],
            [
              0,
              0,
              0,
              1
            ]
          ]
        },
        {
          "body" : {
            "children" : [
              {
                "body" : {
                  "kind" : "shape3D",
                  "primitive" : {
                    "cylinder" : {
                      "bottomRadius" : 0.5,
                      "height" : 5,
                      "segmentCount" : 20,
                      "topRadius" : 0.5
                    }
                  }
                },
                "kind" : "transform",
                "transform" : [
                  [
                    1,
                    0,
                    0,
                    0
                  ],
                  [
                    0,
                    1,
                    0,
                    0
                  ],
                  [
                    0,
                    0,
                    1,
                    0
                  ],
                  [
                    0,
                    0,
                    0,
                    1
                  ]
                ]
              },
              {
                "body" : {
                  "kind" : "shape3D",
                  "primitive" : {
                    "cylinder" : {
                      "bottomRadius" : 1,
                      "height" : 2,
                      "segmentCount" : 41,
                      "topRadius" : 0
                    }
                  }
                },
                "kind" : "transform",
                "transform" : [
                  [
                    1,
                    0,
                    0,
                    -0.001467099408129724
                  ],
                  [
                    0,
                    1,
                    0,
                    0
                  ],
                  [
                    0,
                    0,
                    1,
                    5
                  ],
                  [
                    0,
                    0,
                    0,
                    1
                  ]
                ]
              }
            ],
            "kind" : "lazyUnion"
          },
          "kind" : "transform",
          "transform" : [
            [
              -1,
              0,
              0,
              0
            ],
            [
              0,
              1,
              0,
              0
            ],
            [
              0,
              0,
              1,
              -7
            ],
            [
              0,
              0,
              0,
              1
            ]
          ]
        },
        {
          "body" : {
            "children" : [
              {
                "body" : {
                  "kind" : "shape3D",
                  "primitive" : {
                    "cylinder" : {
                      "bottomRadius" : 0.5,
                      "height" : 5,
                      "segmentCount" : 20,
                      "topRadius" : 0.5
                    }
                  }
                },
                "kind" : "transform",
                "transform" : [
                  [
                    1,
                    0,
                    0,
                    0
                  ],
                  [
                    0,
                    1,
                    0,
                    0
                  ],
                  [
                    0,
                    0,
                    1,
                    0
                  ],
                  [
                    0,
                    0,
                    0,
                    1
                  ]
                ]
              },
              {
                "body" : {
                  "kind" : "shape3D",
                  "primitive" : {
                    "cylinder" : {
                      "bottomRadius" : 1,
                      "height" : 2,
                      "segmentCount" : 41,
                      "topRadius" : 0
                    }
                  }
                },
                "kind" : "transform",
                "transform" : [
                  [
                    1,
                    0,
                    0,
                    -0.001467099408129724
                  ],
                  [
                    0,
                    1,
                    0,
                    0
                  ],
                  [
                    0,
                    0,
                    1,
                    5
                  ],
                  [
                    0,
                    0,
                    0,
                    1
                  ]
                ]
              }
            ],
            "kind" : "lazyUnion"
          },
          "kind" : "transform",
          "transform" : [
            [
              1,
              0,
              0,
              0
            ],
            [
              0,
              1,
              -3.1339616320458145e-16,
              4.949747468305832
            ],
            [
              0,
              2.781780846884627e-16,
              1,
              -4.949747468305834
            ],
            [
              0,
              0,
              0,
              1
            ]
          ]
        },
        {
          "body" : {
            "children" : [
              {
                "body" : {
                  "kind" : "shape3D",
                  "primitive" : {
                    "cylinder" : {
                      "bottomRadius" : 0.5,
                      "height" : 5,
                      "segmentCount" : 20,
                      "topRadius" : 0.5
                    }
                  }
                },
                "kind" : "transform",
                "transform" : [
                  [
                    1,
                    0,
                    0,
                    0
                  ],
                  [
                    0,
                    1,
                    0,
                    0
                  ],
                  [
                    0,
                    0,
                    1,
                    0
                  ],
                  [
                    0,
                    0,
                    0,
                    1
                  ]
                ]
              },
              {
                "body" : {
                  "kind" : "shape3D",
                  "primitive" : {
                    "cylinder" : {
                      "bottomRadius" : 1,
                      "height" : 2,
                      "segmentCount" : 41,
                      "topRadius" : 0
                    }
                  }
                },
                "kind" : "transform",
                "transform" : [
                  [
                    1,
                    0,
                    0,
                    -0.001467099408129724
                  ],
                  [
                    0,
                    1,
                    0,
                    0
                  ],
                  [
                    0,
                    0,
                    1,
                    5
                  ],
                  [
                    0,
                    0,
                    0,
                    1
                  ]
                ]
              }
            ],
            "kind" : "lazyUnion"
          },
          "kind" : "transform",
          "transform" : [
            [
              1,
              0,
              0,
              0
            ],
            [
              0,
              1,
              2.4492935982947064e-16,
              7
            ],
            [
              0,
              -2.4492935982947064e-16,
              1,
              -1.2858791391047208e-15
            ],
            [
              0,
              0,
              0,
              1
            ]
          ]
        },
        {
          "body" : {
            "children" : [
              {
                "body" : {
                  "kind" : "shape3D",
                  "primitive" : {
                    "cylinder" : {
                      "bottomRadius" : 0.5,
                      "height" : 5,
                      "segmentCount" : 20,
                      "topRadius" : 0.5
                    }
                  }
                },
                "kind" : "transform",
                "transform" : [
                  [
                    1,
                    0,
                    0,
                    0
                  ],
                  [
                    0,
                    1,
                    0,
                    0
                  ],
                  [
                    0,
                    0,
                    1,
                    0
                  ],
                  [
                    0,
                    0,
                    0,
                    1
                  ]
                ]
              },
              {
                "body" : {
                  "kind" : "shape3D",
                  "primitive" : {
                    "cylinder" : {
                      "bottomRadius" : 1,
                      "height" : 2,
                      "segmentCount" : 41,
                      "topRadius" : 0
                    }
                  }
                },
                "kind" : "transform",
                "transform" : [
                  [
                    1,
                    0,
                    0,
                    -0.001467099408129724
                  ],
                  [
                    0,
                    1,
                    0,
                    0
                  ],
                  [
                    0,
                    0,
                    1,
                    5
                  ],
                  [
                    0,
                    0,
                    0,
                    1
                  ]
                ]
              }
            ],
            "kind" : "lazyUnion"
          },
          "kind" : "transform",
          "transform" : [
            [
              1,
              0,
              0,
              0
            ],
            [
              0,
              0.9999999999999998,
              1.5638691733620395e-16,
              4.949747468305832
            ],
            [
              0,
              -1.6715578222594702e-16,
              0.9999999999999998,
              4.949747468305832
            ],
            [
              0,
              0,
              0,
              1
            ]
          ]
        }
      ],
      "kind" : "boolean",
      "type" : "union"
    }
  ]
}